precompile/allow_list.go

```diff
@@ -0,0 +1,207 @@
+}
+
+// PackModifyAllowList packs [address] and [role] into the appropriate arguments for modifying the allow list.
+// Note: [role] is not packed in the input value returned, but is instead used as a selector for the function
+// selector that should be encoded in the input.
+func PackModifyAllowList(address common.Address, role AllowListRole) ([]byte, error) {
+	// function selector (4 bytes) + hash for address
+	input := make([]byte, 0, selectorLen+common.HashLength)
+
+	switch role {
+	case AllowListAdmin:
+		input = append(input, setAdminSignature...)
+	case AllowListEnabled:
+		input = append(input, setDeployerSignature...)
+	case AllowListNoRole:
+		input = append(input, setNoneSignature...)
+	default:
+		return nil, fmt.Errorf("cannot pack modify list input with invalid role: %s", role)
+	}
+
+	input = append(input, address.Hash().Bytes()...)
+	return input, nil
+}
+
+// PackReadAllowList packs [address] into the input data to the read allow list function
+func PackReadAllowList(address common.Address) []byte {
+	input := make([]byte, 0, selectorLen+common.HashLength)
+	input = append(input, readAllowListSignature...)
+	input = append(input, address.Hash().Bytes()...)
+	return input
+}
+
+// writeAllowList verifies that [callerAddr] has the correct permissions to modify the allow list and if so updates [modifyAddress] permissions
+// to [role].
+func writeAllowList(evm PrecompileAccessibleState, callerAddr common.Address, modifyAddress common.Address, role AllowListRole, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	// Note: this should never happen since the required gas should be verified before calling Run.
+	if suppliedGas < ModifyAllowListGasCost {
+		return nil, 0, fmt.Errorf("running allow list exceeds gas allowance (%d) < (%d)", ModifyAllowListGasCost, suppliedGas)
+	}
+
+	remainingGas = suppliedGas - ModifyAllowListGasCost
+	if readOnly {
+		return nil, remainingGas, ErrReadOnlyModifyAllowList
+	}
+
+	// Verify that the caller is in the allow list and therefore has the right to modify it
+	callerStatus := GetAllowListStatus(evm.GetStateDB(), callerAddr)
+	if !callerStatus.IsAdmin() {
+		return nil, remainingGas, fmt.Errorf("%w: %s", ErrCannotModifyAllowList, callerAddr)
+	}
+
+	SetAllowListRole(evm.GetStateDB(), modifyAddress, role)
+
+	// Return an empty output and the remaining gas
+	return []byte{}, remainingGas, nil
+}
+
+// createAllowListSetter returns an execution function for setting the allow list status of the input address argument to [role].
+func createAllowListSetter(role AllowListRole) func(evm PrecompileAccessibleState, callerAddr common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	return func(evm PrecompileAccessibleState, callerAddr, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+		if len(input) != common.HashLength {
```

Comment:
Shouldn't this be address length?

Reply:
No this is `common.HashLength` because an address is encoded in a 32 byte slot within the EVM, so in order to provide compatibility with calling via a simple unmodified library contract this needs to be `common.HashLength` and parse the address in the same way.

Reply:
So this is a 32 byte slot filed only with the first 20 bytes?

Reply:
Actually the bytes are shifted over the be the last 20 bytes